[TRB-43185] automation test for dynamic logging

shayne-lin opened the pull request:
# Intent
https://vmturbo.testrail.com/index.php?/cases/view/403081&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=96899

Add automation tests for dynamic log change feature (https://github.com/turbonomic/kubeturbo/pull/900). The test performs the following: 
1. Create ns
2. Install kubeturbo through operator
~~3. Rediscover in case discovery is not initiated in time.~~
4. Check kubeturbo pod to make sure deployment is successful.
5. Change log level by updating CR (spec.logging.level)
6. Look for msg "Logging level is changed from 2 to 5"
7. Check pod doesn't restart and old logs are preserved
8. Check that cm is updated
~~9. Rediscover to generate new logs~~
~~10. Check pod logs again to make sure logs with correct loglevel are showing up.~~

# Background

Details on dynamic log changes. https://jsw.ibm.com/browse/TRB-43185

# Testing

Automation tests results documented in the jira issue above. Here's the log output from the tests:

Logs for test run:
```
OPERATOR_IMAGE_VERSION=8.9.5-SNAPSHOT KUBETURBO_IMAGE_VERSION=8.9.5-SNAPSHOT bash ./test.sh
> Tearing up kubeturbo tests
namespace/turbo created
customresourcedefinition.apiextensions.k8s.io "kubeturbos.charts.helm.k8s.io" deleted
customresourcedefinition.apiextensions.k8s.io/kubeturbos.charts.helm.k8s.io created
serviceaccount/kubeturbo-operator created
clusterrolebinding.rbac.authorization.k8s.io/kubeturbo-operator created
deployment.apps/kubeturbo-operator created
> Start testing for Deploy single kubeturbo test
namespace/testns1 created
kubeturbo.charts.helm.k8s.io/kubeturbo-testcr1 created
Wait for 30s to finish container creation
>  Deployment deployment.apps/kubeturbo-testcr1 cr check........PASSED
>  Deployment deployment.apps/kubeturbo-testcr1 log check........PASSED
>  ClusterRoleBinding turbo-all-binding-kubeturbo-testcr1-testns1 check........PASSED
>  ClusterRole cluster-admin check........PASSED
> Deploy single kubeturbo test........PASSED
> Start testing for Deploy multiple kubeturbos test
namespace/testns2 created
kubeturbo.charts.helm.k8s.io/kubeturbo-testcr2 created
Wait for 30s to finish container creation
>  Deployment deployment.apps/kubeturbo-testcr2 cr check........PASSED
>  Deployment deployment.apps/kubeturbo-testcr2 log check........PASSED
>  ClusterRoleBinding turbo-all-binding-kubeturbo-testcr2-testns2 check........PASSED
>  ClusterRole cluster-admin check........PASSED
> Deploy multiple kubeturbos test........PASSED
> Start testing for Deploy single turbo-cluster-reader kubeturbo test
namespace/testns3 created
kubeturbo.charts.helm.k8s.io/kubeturbo-testcr3 created
Wait for 30s to finish container creation
>  Deployment deployment.apps/kubeturbo-testcr3 cr check........PASSED
>  Deployment deployment.apps/kubeturbo-testcr3 log check........PASSED
>  ClusterRoleBinding turbo-all-binding-kubeturbo-testcr3-testns3 check........PASSED
>  ClusterRole turbo-cluster-reader-kubeturbo-testcr3-testns3 check........PASSED
> Deploy single turbo-cluster-reader kubeturbo test........PASSED
> Start testing for Deploy multiple turbo-cluster-reader kubeturbos test
namespace/testns4 created
kubeturbo.charts.helm.k8s.io/kubeturbo-testcr4 created
Wait for 30s to finish container creation
>  Deployment deployment.apps/kubeturbo-testcr4 cr check........PASSED
>  Deployment deployment.apps/kubeturbo-testcr4 log check........PASSED
>  ClusterRoleBinding turbo-all-binding-kubeturbo-testcr4-testns4 check........PASSED
>  ClusterRole turbo-cluster-reader-kubeturbo-testcr4-testns4 check........PASSED
> Deploy multiple turbo-cluster-reader kubeturbos test........PASSED
> Tearing down kubeturbo tests
kubeturbo.charts.helm.k8s.io "kubeturbo-testcr1" deleted
namespace "testns1" deleted
kubeturbo.charts.helm.k8s.io "kubeturbo-testcr2" deleted
namespace "testns2" deleted
kubeturbo.charts.helm.k8s.io "kubeturbo-testcr3" deleted
namespace "testns3" deleted
kubeturbo.charts.helm.k8s.io "kubeturbo-testcr4" deleted
namespace "testns4" deleted
deployment.apps "kubeturbo-operator" deleted
clusterrolebinding.rbac.authorization.k8s.io "kubeturbo-operator" deleted
serviceaccount "kubeturbo-operator" deleted
namespace "turbo" deleted
Test passed!
> Running dynamic logging tests
namespace/turbo created
customresourcedefinition.apiextensions.k8s.io "kubeturbos.charts.helm.k8s.io" deleted
customresourcedefinition.apiextensions.k8s.io/kubeturbos.charts.helm.k8s.io created
serviceaccount/kubeturbo-operator created
clusterrolebinding.rbac.authorization.k8s.io/kubeturbo-operator created
deployment.apps/kubeturbo-operator created
> Start testing for Deploy kubeturbo for dynamic logging
namespace/testns1 created
kubeturbo.charts.helm.k8s.io/kubeturbo-testcr1 created
Wait for 30s to finish container creation
>  Deployment deployment.apps/kubeturbo-testcr1 cr check........PASSED
>  Deployment deployment.apps/kubeturbo-testcr1 log check........PASSED
>  ClusterRoleBinding turbo-all-binding-kubeturbo-testcr1-testns1 check........PASSED
>  ClusterRole cluster-admin check........PASSED
> Deploy kubeturbo for dynamic logging........PASSED
Updating CR to add logging level...
kubeturbo.charts.helm.k8s.io/kubeturbo-testcr1 configured
Wait for 10s for log level changes to be reflected...
Wait for 10s for log level changes to be reflected...
Wait for 10s for log level changes to be reflected...
Wait for 10s for log level changes to be reflected...
I0725 16:38:46.720125 1 kubeturbo_builder.go:1045] Logging level is changed from 2 to 5
Cleanup...
kubeturbo.charts.helm.k8s.io "kubeturbo-testcr1" deleted
namespace "testns1" deleted
deployment.apps "kubeturbo-operator" deleted
clusterrolebinding.rbac.authorization.k8s.io "kubeturbo-operator" deleted
serviceaccount "kubeturbo-operator" deleted
namespace "turbo" deleted
Test passed!

```

```

# Checklist

These are the items that must be done by the developer and by reviewers before the change is ready to merge. Please ~~strikeout~~ any items that are not applicable, but don't delete them

- [ ] Developer Checks
    - [ ] Full build with unit tests and fmt and vet checks
    - [ ] Unit tests added / updated
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Integration tests added / updated
    - [ ] Manual testing done (and described)
    - [ ] Product sweep run and passed
    - [ ] Developer wiki updated (and linked to this description)
- [ ] Reviewer Checks
    - [ ] Merge request description clear and understandable
    - [ ] Developer checklist items complete
    - [ ] Functional code review (how is the code written)
    - [ ] Architectural review (does the code try to do the right thing, in the right way)
    - [ ] Defensive coding (incoming data checked / sanitized, exceptions logged, clear error messages)
    - [ ] No unlicensed images, no third-party code (such as from StackOverflow)
    - [ ] Security review checklist complete.

# Audience

_(@ mention any `review/...` groups or people that should be aware of this merge request)_

kevinwangcn commented on deploy/kubeturbo-operator/test.sh:
```diff
@@ -2,48 +2,56 @@
+WAIT_FOR_DEPLOYMENT=30
 OPERATOR_IMAGE_VERSION="8.9.5-SNAPSHOT"
 OPERATOR_IMAGE="icr.io\/cpopen\/kubeturbo-operator:${OPERATOR_IMAGE_VERSION}"
+KUBETURBO_IMAGE_VERSION="8.9.5-SNAPSHOT"
```
I don't think it's right to hardcode this version. Hardcoding the version means we always test this `8.9.5` version. It loses the point of the testing.  This version should be configurable or passed from the command line.

shayne-lin replied:
Updated PR to use the existing VERSION env var instead of hardcoding it in test.sh. You can now overwrite the specific version like this: `make test VERSION=8.9.5-SNAPSHOT`

LionelWuW replied:
In my other PR, I read the version and the image registry from the environment so that the Jenkins build can pass the version to the script. While the hardcoded version can be treated as the safeguard to make sure the script passes with a certain version.

kevinwangcn commented on deploy/kubeturbo-operator/test.sh:
```diff
@@ -2,48 +2,56 @@
 
+function rediscover_target {
+	DISPLAY_NAME=$1
+    ip=$($KUBECTL get services --no-headers -n $namespace | grep nginx | awk '{print $4; exit}' | cut -d "," -f 1)
```
I don't quite understand here. Why are we trying to grep `nginx` service?  What it it used for? who will deploy it?

shayne-lin replied:
It's just one of the ways to detect the IP of the instance where I ran my tests instead of hardcoding it. It might need to be updated when @LionelWuW integrates the tests in jenkins.

LionelWuW replied:
This is the same way that the exported method `xl_version` uses in the `.bashrc` to get the IP address.

kevinwangcn replied:
I understand you're trying to make a broadcast through`nginx` here. But If we have this in here, it means our test will strongly depend on the server component.  Kubeturbo has to be deployed in the same namespace as the other server component. Usually, that's not our deploy model except turbo on turbo.  Moreover, I'm not sure if the DevOps team can help with that when they run this test in the Jenkins pipeline.

shayne-lin replied:
I am using to the target API to rediscover the target, so this method will depend on the server regardless. Is there another way to trigger rediscovery of target without going through the server? 

btw, for the testcases, we are using deploying kubeturbo operator and kubeturbo in separate ns than the server component.

kevinwangcn replied:
I'm thinking about another way to verify this instead of a broadcast. What I have in my mind is we can kill the kubeturbo which will trigger the following message.  If we take it this way, I guess we don't need to involve a server component.  Does this make sense?
```
glog.V(4).Infof("*** Handling Kubeturbo Termination ***")
```

shayne-lin replied:
After discussion, I removed the rediscover logic so there's no dependency on XL server.

kevinwangcn commented on deploy/kubeturbo-operator/test.sh:
```diff
@@ -161,15 +174,47 @@ function uninstall_cr {
+		echo "Wait for 10s for log level changes to be reflected..."
+		sleep 10
+		MSG=$($KUBECTL -n ${NAMESPACE} logs ${POD_NAME} | grep "Logging level is changed from")
+		if [ $COUNT -eq 10 ]; then
```
Maybe changing `10` to `15` is more safe.

shayne-lin replied:
I am actually doing a loop that checks every 10 secs, so it will wait until that line of log shows up, or time out when it loops 10 times (i.e. 100 seconds).

kevinwangcn replied:
Yes, I know you're doing a 100s waiting here.  I suggested to make a little bigger.  From my experience, sometimes it nearly takes 2mins

shayne-lin replied:
I see. Usually for me it takes around 40 to 50 secs, but I increased the timeout to 15 just to be safe.

kevinwangcn commented:
@shayne-lin  thanks for doing this.  Usually, we fork Kubeturbo to our personal GitHub repository and create PR from there. We don't create the branch directly in the Turbonomic repository. And for the commit,  we always put the JIRA number in the comment. Please check other PRs for reference.

shayne-lin commented:
> @shayne-lin thanks for doing this. Usually, we fork Kubeturbo to our personal GitHub repository and create PR from there. We don't create the branch directly in the Turbonomic repository. And for the commit, we always put the JIRA number in the comment. Please check other PRs for reference.

Sounds good. I added JIRA issue in the commit and updated the PR title. Will fork next time when making code changes.

shayne-lin commented:
Closing the PR as the tests will be moved to core-python.
